src/lib/scheduler/worker.cpp

```diff
@@ -128,17 +140,59 @@ void Worker::join() {
 uint64_t Worker::num_finished_tasks() const { return _num_finished_tasks; }
 
 void Worker::_wait_for_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) {
-  auto tasks_completed = [&tasks]() {
-    // Reversely iterate through the list of tasks, because unfinished tasks are likely at the end of the list.
-    for (auto it = tasks.rbegin(); it != tasks.rend(); ++it) {
-      if (!(*it)->is_done()) {
+  // This lambda checks if all tasks from the vector have been executed. If they are, it causes _wait_for_tasks to
+  // return. If there are remaining tasks, it primarily tries to execute these. If they cannot be executed, the
+  // worker performs work for others (i.e., executes tasks from the queue).
+  auto check_own_tasks = [&]() {
+    auto all_done = true;
+    // Note: If we found a task that has not yet been executed, we reset this loop and start from the beginning.
+    // As such, both all_done and it may be reset outside of the following line.
+    for (auto it = tasks.begin(); it != tasks.end(); ++it) {
+      const auto& task = *it;
+      if (task->is_done()) {
+        continue;
+      }
+
+      // Task is not yet done - check if it is ready for execution
+      if (!task->is_ready()) {
+        all_done = false;
+        continue;
+      }
+
+      // Give other tasks, i.e., tasks that are unrelated to what we are currently waiting on, a certain chance of being
+      // executed, too. Anectotal evidence says that this is a good idea. For some reason, this keeps the memory
+      // consumption of TPC-H Q6 low even if the scheduler is overcommitted. Because generating random numbers is
+      // somewhat expensive, we keep a list of random numbers and reuse them.
+      // TODO(anyone): Look deeper into scheduling theory and make this theoretically sound.
+      _next_random = (_next_random + 1) % _random.size();
+      if (_random[_next_random] <= 20) {
         return false;
       }
+
+      // Run one of our own tasks. First, let everyone know that we are about to execute it. This is necessary because
+      // the task is already in a queue and some other worker might pull it at the same time.
+      const auto successfully_assigned = task->try_mark_as_assigned_to_worker();
+      if (!successfully_assigned) {
+        // Some other worker has already started to work on this task - pick a different one.
+        all_done = false;
+        continue;
+      }
+
+      // Actually execute it.
+      task->execute();
+      ++_num_finished_tasks;
+
+      // Reset loop so that we re-visit tasks that may have finished in the meantime. We need to decrement `it` because
+      // it will be incremented when the loop iteration finishes.
+      all_done = true;
+      it = tasks.begin() - 1;
```

Comment:
Isn't that UB? https://timsong-cpp.github.io/cppwp/n4659/iterator.requirements#bidirectional.iterators

We don't reference it, but I am not sure we can rely on the STL doing what you expect it to.

Reply:
Argh, I remember having that conversation.

...

Outer join!

I refactored this loop anyway, please have another look.